substrate/frame/balances/src/lib.rs

```diff
@@ -544,6 +550,19 @@ pub mod pallet {
 				!<T as Config<I>>::ExistentialDeposit::get().is_zero(),
 				"The existential deposit must be greater than zero!"
 			);
+
+			assert!(
+				T::MaxHolds::get() >= <T::RuntimeHoldReason as VariantCount>::variant_count(),
+				"MaxHolds should be greater than or equal to the number of hold reasons: {} < {}",
+				T::MaxHolds::get(),
+				<T::RuntimeHoldReason as VariantCount>::variant_count()
+			);
```

Comment:
But there can only be `variant_count` holds on any account or? So, we actually don't need `T::MaxHolds`?

Reply:
The `MaxHolds` and `MaxFreezes` are used as bounds for `BoundedVec`s, so we need to be careful about changing them.  
If we make it implicit then it could become undecodable when one of the reasons is removed.

Reply:
If you remove a reason, that is stored in state, it will become undecodable any way :P

Reply:
Good point, in a more cowboy world we could have just have no `MaxHolds` and just use `variant_count`. 

As long as you only add variants, it is fine. But as you start removing, you are going to destroy yourself. 

In the current system, you can kinda avoid this by setting a large `MaxHold` and hoping it will never be reached. 

But I agree that the real solution to me, IMO, is to have a system that reads the previous runtime at compile time (ie use `subxt`), then make sure in a new runtime these `MaxBlahBlah`s are not decremented recklessly. This would be feature gated, so you won't always do it. I formulated this in the past in #244.

Reply:
In practice, `MaxHolds` is implemented as a `const` even though it's exposed why non-const `Get` trait.
If it is to be exposed via `const` in addition to that, you have a `const_assert!` in your runtime so it's checked at compile time for every runtime that has this pallet and no additional test is required (that might not be run by others).

Reply:
> As long as you only add variants, it is fine. But as you start removing, you are going to destroy yourself.

You would already "destroy" yourself by removing a variant, because that would make the decoding fail as well.

Reply:
There is no real difference between the `BoundedVec` will reject because there are too many elements and the `BoundedVec` will fail to decode there  is an unknown variant in it.

Reply:
Opened https://github.com/paritytech/polkadot-sdk/issues/2000 as follow-up.